### DISCUSSION TITLE
Is this a problem with the accuracy of the positioning system or a communication problem?

### DISCUSSION THREAD

**Discussion by swarm5** [USC-ACTLab/crazyswarm#333] [Q&A]
When I was doing a multi-cfs flight test, my terminal printed a lot of error messages. I would like to ask, is this related to the radio channel or the accuracy of the positioning system?
1. When I only use single point tracking:
**①What is the warning of "No updated pose for CF'"? It's strange that all cfs exist. 0
There is no update data for .07s.
②What does the warning ‘’all dynamic check failed for object ‘’ mean?**
③Sometimes it still exists, and all cf tracking is skipped. Which is ‘Lost tracking for object cf1 skipping’
> [ WARN] [1614006672.979794508]: No updated pose for CF cf1 for 0.070621 s.
> [ WARN] [1614006672.979817155]: No updated pose for CF cf4 for 0.070621 s.
> [ WARN] [1614006672.979834739]: No updated pose for CF cf7 for 0.070621 s.
> [ WARN] [1614006672.979850688]: No updated pose for CF cf10 for 0.070621 s.
> [ WARN] [1614006672.979866029]: No updated pose for CF cf13 for 0.070621 s.
> [ WARN] [1614006672.979880373]: No updated pose for CF cf16 for 0.070621 s.
> [ WARN] [1614006672.979899806]: No updated pose for CF cf19 for 0.070621 s.
> [ WARN] [1614006672.979926160]: No updated pose for CF cf2 for 0.070633 s.
> [ WARN] [1614006672.979947995]: No updated pose for CF cf5 for 0.070633 s.
> [ WARN] [1614006672.979961669]: No updated pose for CF cf8 for 0.070633 s.
> [ WARN] [1614006672.979980356]: No updated pose for CF cf11 for 0.070633 s.
> [ WARN] [1614006672.979996091]: No updated pose for CF cf14 for 0.070633 s.
> [ WARN] [1614006672.980008042]: No updated pose for CF cf17 for 0.070633 s.
> [ WARN] [1614006672.980019598]: No updated pose for CF cf20 for 0.070633 s.
> [ WARN] [1614006672.980064612]: all dynamic check failed for object cf12
> [ WARN] [1614006672.980094394]: all dynamic check failed for object cf15
> [ WARN] [1614006672.980113381]: all dynamic check failed for object cf18
> [ WARN] [1614006672.987918019]: all dynamic check failed for object cf3
> [ WARN] [1614006672.987960247]: all dynamic check failed for object cf6
> [ WARN] [1614006672.987979310]: all dynamic check failed for object cf9
> [ WARN] [1614006672.988013835]: all dynamic check failed for object cf1
> [ WARN] [1614006672.988053225]: all dynamic check failed for object cf4
> [ WARN] [1614006672.988074215]: all dynamic check failed for object cf7
> [ WARN] [1614006672.988099212]: all dynamic check failed for object cf10
> [ WARN] [1614006672.988123468]: all dynamic check failed for object cf13
> [ WARN] [1614006672.988159828]: all dynamic check failed for object cf2
> [ WARN] [1614006672.988203386]: all dynamic check failed for object cf5
> [ WARN] [1614006672.988225352]: all dynamic check failed for object cf8
> [ WARN] [1614006672.988258853]: all dynamic check failed for object cf14
> [ WARN] [1614006672.988279019]: all dynamic check failed for object cf17
> [ WARN] [1614006672.988301216]: all dynamic check failed for object cf20

2. When I use four markers with the same configuration:
**Four markers also exist without pose update and Dynamic check failed for object. What does this warning mean?**
> [ WARN] [1614248582.617904574]: No updated pose for CF cf1 for 0.538768 s.
> [pcl::IterativeClosestPoint::computeTransformation] Not enough correspondences found. Relax your threshold parameters.
> [ WARN] [1614248591.623150055]: ICP did not converge! for object cf2
> [ WARN] [1614248591.623284223]: No updated pose for CF cf2 for 0.000956 s.
> [ WARN] [1614250536.203208257]: No updated pose for CF cf5 for 0.066820 s.
> [ WARN] [1614250536.208343736]: Dynamic check failed for object cf5
> wroll: 34.3777 >= 25
> wyaw: -21.0298 >= 15
> roll: 2.67076 >= 1.4
> fitness: 0.00228335 >= 0.001
> [ WARN] [1614250536.208474209]: No updated pose for CF cf5 for 0.072178 s.
> [ WARN] [1614250536.214027932]: Dynamic check failed for object cf5
> wyaw: 34.0847 >= 15
> fitness: 0.00119523 >= 0.001
> [ WARN] [1614250536.214159560]: No updated pose for CF cf5 for 0.077857 s.
> [ WARN] [1614250536.219347543]: Dynamic check failed for object cf5
> wyaw: 18.5341 >= 15
> fitness: 0.00124195 >= 0.001
> [ WARN] [1614250536.219479344]: No updated pose for CF cf5 for 0.083188 s.
> [ WARN] [1614250536.225068322]: Dynamic check failed for object cf5
> wyaw: 19.5648 >= 15
> roll: 2.34783 >= 1.4
> fitness: 0.00123962 >= 0.001
> [ WARN] [1614250536.225184946]: No updated pose for CF cf5 for 0.088915 s.
> [ WARN] [1614250536.230637103]: Dynamic check failed for object cf5
> wyaw: -31.1749 >= 15
> [ WARN] [1614250536.230772946]: No updated pose for CF cf5 for 0.094482 s.
> [ WARN] [1614250536.236178660]: Dynamic check failed for object cf5
> wyaw: -15.1722 >= 15
> roll: 2.67076 >= 1.4
> fitness: 0.00228335 >= 0.001
> [ WARN] [1614250536.203208257]: No updated pose for CF cf5 for 0.066820 s.
> [ WARN] [1614250536.208343736]: Dynamic check failed for object cf5
> wroll: 34.3777 >= 25
> wyaw: -21.0298 >= 15
> roll: 2.67076 >= 1.4
> fitness: 0.00228335 >= 0.001
> 
> [ WARN] [1614250536.208474209]: No updated pose for CF cf5 for 0.072178 s.
> [ WARN] [1614250536.214027932]: Dynamic check failed for object cf5
> wyaw: 34.0847 >= 15
> fitness: 0.00119523 >= 0.001
> 
> [ WARN] [1614250536.214159560]: No updated pose for CF cf5 for 0.077857 s.
> [ WARN] [1614250536.219347543]: Dynamic check failed for object cf5
> wyaw: 18.5341 >= 15
> fitness: 0.00124195 >= 0.001
> 
> [ WARN] [1614250536.219479344]: No updated pose for CF cf5 for 0.083188 s.
> [ WARN] [1614250536.225068322]: Dynamic check failed for object cf5
> wyaw: 19.5648 >= 15
> roll: 2.34783 >= 1.4
> fitness: 0.00123962 >= 0.001
> 
> [ WARN] [1614250536.225184946]: No updated pose for CF cf5 for 0.088915 s.
> [ WARN] [1614250536.230637103]: Dynamic check failed for object cf5
> wyaw: -31.1749 >= 15
> 
> [ WARN] [1614250536.230772946]: No updated pose for CF cf5 for 0.094482 s.
> [ WARN] [1614250536.236178660]: Dynamic check failed for object cf5
> wyaw: -15.1722 >= 15
> roll: 2.67076 >= 1.4
> fitness: 0.00228335 >= 0.001
> 
**3.Dynamic check failed for objectWarning, can it be eliminated by enlarging dynamicsConfigurations in crazyflieTypes.yaml?**
**4. My own guess is that "Dynamic check failed for object" is more due to the limited accuracy of the motion capture system, while "No updated pose for CF" may be due to communication problems, and there is another possibility that is due to "Dynamic check failed for object "caused to be unable to track for a long time**

**Comment by whoenig** [comment]
All those messages are related to your mocap and have nothing to do with communication with the Crazyflie itself.

* No updated pose for <x> time means that there was likely no pointcloud for a couple of frames. In the current design, this can also happen at start-up, because the main communication loop is slowed down such that mocap updates are not processed fast enough. If you see those messages during the flight, this indicates a problem with either your mocap or your PC being too slow.
* The dynamics check is a very basic outlier filter that removes estimates that are physically impossible during normal flight. This usually indicates bad mocap coverage, especially for the 4-marker tracking mode. You can adjust the configuration in crazyflieTypes.yaml to allow larger limits, but this usually is counterproductive and increases the number of crashes.

**Reply by jpreiss** [reply]
Maybe we should suppress those warnings during startup. They (understandably) confuse a lot of users.

**Reply by whoenig** [reply]
It's hard to suppress just at start-up. The proper (long-term) solution is to switch to [crazyflie-link-cpp](https://github.com/bitcraze/crazyflie-link-cpp) in the backend, which will remove those bottlenecks and allow features like uploading trajectories during the flight.

**Reply by swarm5** [reply]
Thank you for your answer, I am not sure what function crazyflie-link-cpp has? How to switch to crazyflie-link-cpp in the background.
**However, I did a test in the afternoon. Using the latest firmware version, I can upload the trajectory during the flight. I have already tested it many times.**

**Reply by whoenig** [reply]
This comment was for @jpreiss. You cannot use crazyflie-link-cpp yet. It is not guaranteed that uploading works while flying - you might have been just lucky.
